### DISCUSSION TITLE
構文エラー

### DISCUSSION THREAD

**Discussion by hayashi-ay** [katataku/minishell#15] [General]
構文解析時にエラーにするもの。

- 閉じられてないクオート（ダブル、シングル、バック）
- ~~セミコロン、バックスラッシュ~~ 通常の文字として解釈する
- ~~＆~~ 通常の文字として解釈する
- 閉じられていない括弧(カーリーブラケット{})

**Comment by hayashi-ay** [comment]
構文エラー時の終了ステータスは？

**Reply by katataku** [reply]
以下の実行結果に従うと2ですかね

```
$ cat ./test.sh
cat "aa
$ bash ./test.sh
./test.sh: line 1: unexpected EOF while looking for matching `"'
./test.sh: line 2: syntax error: unexpected end of file
$ echo $?       
2
```

**Reply by katataku** [reply]
https://tldp.org/LDP/abs/html/exitcodes.html#EXITCODESREF
これが出典かも？

**Reply by katataku** [reply]
実装しようとして気づきました。これ終了ステータスが2ってことはlexer内でexit(2)を呼び出すってことですよね？
ってことは、lexerも子プロセスとして実行しないといけない！？

exitせずに終了ステータス入れる方法あるのかなー

**Reply by katataku** [reply]
これができるってことはあるはず！？


```bash
bash-3.2$ cat test.sh
cd "hoge
bash-3.2$ bash ./test.sh
./test.sh: line 1: unexpected EOF while looking for matching `"'
./test.sh: line 2: syntax error: unexpected end of file
bash-3.2$ echo $?
2
```

**Reply by hayashi-ay** [reply]
bashの内部変数として直前のコマンドの終了ステータスを格納するものがいるはずで、その変数に値をセットしてあげればいいのかなと

**Reply by katataku** [reply]
なるほどー
やっぱり本当にプロセスの終了ステータス使うわけじゃないんですね。納得です。

**Reply by hayashi-ay** [reply]
シェルの内部変数はグローバル変数になりそうですね〜。

**Reply by katataku** [reply]
ですね！

**Comment by katataku** [comment]
構文解釈時のシンタックスチェックでやること

- <.>,<<,>>の後にはwordが続いていること
- |の後にはwordが続いていること
- { word } という順番であること
- $ word もしくは$ { word } という順番であること
- }の前には{が存在すること
